iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+    auto dialectAttr = LLVM::AsmDialectAttr::get(rewriter.getContext(),
+                                                 LLVM::AsmDialect::AD_ATT);
+    LLVM::InlineAsmOp asmOp = rewriter.create<LLVM::InlineAsmOp>(
+        loc, returnType, asmOperands,
+        R"ASM(
+            sdot $0.4s, $18.16b, $16.4b[0]
+            sdot $1.4s, $19.16b, $16.4b[0]
+            sdot $2.4s, $18.16b, $16.4b[1]
+            sdot $3.4s, $19.16b, $16.4b[1]
+            sdot $4.4s, $18.16b, $16.4b[2]
+            sdot $5.4s, $19.16b, $16.4b[2]
+            sdot $6.4s, $18.16b, $16.4b[3]
+            sdot $7.4s, $19.16b, $16.4b[3]
+            sdot $8.4s, $18.16b, $17.4b[0]
+            sdot $9.4s, $19.16b, $17.4b[0]
+            sdot $10.4s, $18.16b, $17.4b[1]
+            sdot $11.4s, $19.16b, $17.4b[1]
+            sdot $12.4s, $18.16b, $17.4b[2]
+            sdot $13.4s, $19.16b, $17.4b[2]
+            sdot $14.4s, $18.16b, $17.4b[3]
+            sdot $15.4s, $19.16b, $17.4b[3]
+          )ASM",
+        "=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,=w,w,w,w,w,0,1,2,3,4,5,6,"
+        "7,8,9,10,11,12,13,14,15",
+        /* has_side_effects */ false, /* is_align_stack */ false, dialectAttr);
+
+    // Extract result vectors from the asm op.
+    SmallVector<Value, 16> resVec;
+    for (int i = 0; i < 16; ++i) {
+      resVec.push_back(rewriter.create<LLVM::ExtractValueOp>(
+          loc, int32x4Type, asmOp.res(), rewriter.getI64ArrayAttr({i})));
+    }
+
+    // End of the code that's specific to inline assembly. An intrinsics code
+    // path would merge here.
+
+    // Insert the result vectors of size 4 into the overall result vector of
+    // size 64, still 1D.
+    VectorType int32x64xType = VectorType::get({64}, I32Type);
+    Value result = rewriter.create<arith::ConstantOp>(
+        loc, int32x64xType, DenseIntElementsAttr::get(int32x64xType, 0));
+    for (int i = 0; i < 16; ++i) {
+      result = rewriter.create<vector::InsertStridedSliceOp>(
+          loc, resVec[i], result, std::array<int64_t, 1>{4 * i},
+          std::array<int64_t, 1>{1});
+    }
+
+    // Cast the result from 1D to 2D and replace the original vector.contract.
+    VectorType int32x8x8xType = VectorType::get({8, 8}, I32Type);
+    rewriter.replaceOpWithNewOp<vector::ShapeCastOp>(contractionOp,
+                                                     int32x8x8xType, result);
+    return success();
+  }
+};
+
+}  // namespace
+
+namespace {
+struct VectorContractCustomKernelsPass
+    : VectorContractCustomKernelsBase<VectorContractCustomKernelsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<vector::VectorDialect, LLVM::LLVMDialect>();
+  }
+  void runOnOperation() override;
+};
+}  // namespace
+
+void populateVectorContractCustomKernelsPatterns(
+    OwningRewritePatternList &patterns) {
+  MLIRContext *context = patterns.getContext();
+  const auto &targetOptions = IREE::HAL::getLLVMTargetOptionsFromFlags();
```

Comment:
Calling this from here is too heavy weight. There is a different way to propogate the information you need here. You first add a field to the config struct here https://github.com/google/iree/blob/4027ac99d97b7f11e7a9b8435773fb91909e631a/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp#L705.

That gets propogated as an attribute on the `hal.executable.variant`. You can access the config using the approach here  https://github.com/google/iree/blob/4027ac99d97b7f11e7a9b8435773fb91909e631a/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L70

(Adding a utility method for the above would be good too!).

I would suggest just adding a boolean that if set represents `dotprod` being available.

Reply:
Thank you for the suggestion. Implemented. I kept a minimal data structure `CustomKernelsTargetInfo` for a reason explained in the comment on it, but switched to populating it from the parent HAL::ExecutableVariantOp as you explained, or, in lit tests, from custom options to my pass.